lib/ansible/modules/windows/win_firewall_rule.ps1

```diff
@@ -1,273 +1,164 @@
-# along with Ansible. If not, see <http://www.gnu.org/licenses/>.
-#
+# Copyright (c) 2017 Artem Zinenko <zinenkoartem@gmail.com>
+# Copyright (c) 2014 Timothy Vandenbrande <timothy.vandenbrande@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
 # WANT_JSON
 # POWERSHELL_COMMON
 
-# TODO: Reimplement this using Powershell cmdlets
+function Parse-ProtocolType {
+    param($protocol)
 
-$ErrorActionPreference = "Stop"
+    $protocolNumber = $protocol -as [int]
+    if ($protocolNumber -is [int]) {
+        return $protocolNumber
+    }
 
-function convertToNetmask($maskLength) {
-    [IPAddress] $ip = 0
-    $ip.Address = ([UInt32]::MaxValue) -shl (32 - $maskLength) -shr (32 - $maskLength)
-    return $ip.IPAddressToString
+    switch -wildcard ($protocol) {
+        "tcp" { return [System.Net.Sockets.ProtocolType]::Tcp -as [int] }
+        "udp" { return [System.Net.Sockets.ProtocolType]::Udp -as [int] }
+        "icmpv4*" { return [System.Net.Sockets.ProtocolType]::Icmp -as [int] }
+        "icmpv6*" { return [System.Net.Sockets.ProtocolType]::IcmpV6 -as [int] }
+        default { throw "Unknown protocol '$protocol'." }
+    }
 }
 
-function ConvertTo-TitleCase($string) {
-    return (Get-Culture).TextInfo.ToTitleCase($string.ToLower())
-}
+# See 'Direction' constants here: https://msdn.microsoft.com/en-us/library/windows/desktop/aa364724(v=vs.85).aspx
+function Parse-Direction {
+    param($directionStr)
 
-function ConvertTo-SortedKV($object, $unsupported = @()) {
-    $output = ""
-    foreach($item in $object.GetEnumerator() | Sort -Property Name) {
-        if (($item.Name -notin $unsupported) -and ($item.Value -ne $null)) {
-            $output += "$($item.Name): $($item.Value)`n"
-        }
+    switch ($directionStr) {
+        "in" { return 1 }
+        "out" { return 2 }
+        default { throw "Unknown direction '$directionStr'." }
     }
-    return $output
 }
 
-function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
-    if ($key -eq 'RemoteIP') {
-        if ($outputValue -eq $fwsettingValue) {
-            return $true
-        }
+# See 'Action' constants here: https://msdn.microsoft.com/en-us/library/windows/desktop/aa364724(v=vs.85).aspx
+function Parse-Action {
+    param($actionStr)
 
-        if ($outputValue -eq $fwsettingValue+'-'+$fwsettingValue) {
-            return $true
-        }
+    switch ($actionStr) {
+        "block" { return 0 }
+        "allow" { return 1 }
+        default { throw "Unknown action '$actionStr'." }
+    }
+}
 
-        if (($outputValue -eq $fwsettingValue+'/32') -or ($outputValue -eq $fwsettingValue+'/255.255.255.255')) {
-            return $true
+# Profile enum values: https://msdn.microsoft.com/en-us/library/windows/desktop/aa366303(v=vs.85).aspx
+function Parse-Profiles
+{
+    param($profilesStr)
+
+    $profiles = ($profilesStr.Split(',') | Select -uniq | ForEach {
```

Comment:
Instead of splitting the string here, would be better to change the type of `$profiles` with `Get-AnsibleParam` to list. It will automatically convert strings with `,` separators but also supports people passing in a list themselves.